lib/b3fs/src/bucket/dir/state/mod.rs

```diff
@@ -0,0 +1,314 @@
+
+        // A valid file name never contains the "null" byte. So we can use it as terminator
+        // instead of prefixing the length.
+        buffer.put_u8(0x00);
+
+        // Fill the rest of the bytes with the content link.
+        buffer.extend_from_slice(content);
+        if is_sym {
+            buffer.put_u8(0x00);
+        }
+        self.file.write_all(&buffer).await?;
+
+        Ok(())
+    }
+}
+
+/// Represents the internal state of a directory.
+pub(super) struct InnerDirState<T> {
+    bucket: Bucket,
+    phf_generator: PhfGenerator,
+    next_position: u32,
+    temp_file_path: PathBuf,
+    header_file: HeaderFile,
+    bloom_filter: BloomFilter,
+    pub collector: T,
+}
+
+/// Trait for types that can collect and process directory entries.
+pub trait WithCollector {
+    /// Processes a new entry being inserted into the directory.
+    async fn on_insert(
+        &mut self,
+        borrowed_entry: BorrowedEntry<'_>,
+    ) -> Result<(), errors::InsertError>;
+
+    /// Finalizes the collection process and returns the root hash and hash tree.
+    async fn on_commit(self) -> Result<([u8; 32], Vec<[u8; 32]>), errors::CommitError>;
+}
+
+/// Trait defining the operations that can be performed on a directory state.
+pub trait DirState {
+    /// Inserts a new entry into the directory.
+    async fn insert_entry<'a>(
+        &mut self,
+        borrowed_entry: BorrowedEntry<'a>,
+    ) -> Result<(), errors::InsertError>;
+
+    /// Commits the changes made to the directory.
+    async fn commit(self) -> Result<[u8; 32], errors::CommitError>;
+
+    /// Rolls back any changes made to the directory.
+    async fn rollback(self) -> Result<(), io::Error>;
+}
+
+impl<T: WithCollector> DirState for InnerDirState<T> {
+    async fn insert_entry<'b>(
+        &mut self,
+        borrowed_entry: BorrowedEntry<'b>,
+    ) -> Result<(), errors::InsertError> {
+        let mut bytes = match borrowed_entry.link {
+            BorrowedLink::Content(content) => content,
+            BorrowedLink::Path(path) => path,
+        };
+        let i = self.next_position;
+        let pos = unsafe { NonZeroU32::new_unchecked(i + 1) };
+        self.phf_generator.push(bytes, pos);
+        self.next_position += borrowed_entry.name.len() as u32;
+        self.header_file.insert_entry(borrowed_entry).await?;
+        self.bloom_filter.add(borrowed_entry.name);
+        self.collector.on_insert(borrowed_entry).await?;
+        Ok(())
+    }
+
+    async fn commit(mut self) -> Result<[u8; 32], errors::CommitError> {
+        let (root_hash, tree) = self.collector.on_commit().await?;
+        let hash_tree = self.phf_generator.finalize();
+        self.header_file
+            .commit(&self.bucket, &root_hash, tree, self.bloom_filter, hash_tree)
+            .await?;
+
+        Ok(root_hash)
+    }
+
+    async fn rollback(self) -> Result<(), io::Error> {
+        tokio::fs::remove_dir_all(self.temp_file_path).await
```

Comment:
perhaps this should be `tokio::fs::remove_dir_all`?

Reply:
But it is that, isnt it? I am not following the comment....

Reply:
oh sorry i meant `tokio::fs::remove_file`

Reply:
Take into consideration that this is a random directory inside `wal` dir. Something like `/var/folders/zr/lff92kqx17vb80fs9jjwgd540000gn/T/b3fs/wal/0b172a64fd30f300b1149ca82cb814abfaaac784812969726d30a15aa40866f3/`. Therefore, we are deleting that random dir where we are putting all the transient data for that Dir or File lifetime session. I think it is ok to delete that dir. We are not deleting `wal`